test/functional/rpc_net.py

```diff
@@ -262,6 +263,12 @@ def test_addpeeraddress(self):
         assert_equal(node.addpeeraddress(address="1.2.3.4", port=8333), {"success": False})
         assert_equal(len(node.getnodeaddresses(count=0)), 1)
 
+        self.log.debug("Test adding a second address to the tried table")
+        assert_equal(node.addpeeraddress(address="1.2.3.0", port=8333, tried=True), {"success": True})
+        with node.assert_debug_log(expected_msgs=["Addrman checks started: new 1, tried 1, total 2"]):
+            addrs = node.getnodeaddresses(count=0)  # getnodeaddresses re-runs the addrman checks
+            assert_equal(len(addrs), 2)
```

Comment:
I'm mildly worried this assert could be racy. Saw it fail once with 2 != 1 today.

Reply:
`1.2.3.0` is the same /16 as `1.2.3.4`, and when using `addpeeraddress()`, the source address is set to the destination address. That means that `1.2.3.0` will be placed in the same new bucket as `1.2.3.4`, and has a 1/64 chance of the bucket position colliding (depending on what addrman's nKey has been set to). If that happens, then `1.2.3.0` won't be added to addrman.

You can reduce the chance of a collision by using an address from a different /16.

Reply:
Excellent, thank you! Updating.

Reply:
I think this may still fail with a probability of 1/2^16 = 1/65536 since there could still be bucket/position collisions.

Reply:
Sounds good. WDYT:
```diff
diff --git a/src/rpc/net.cpp b/src/rpc/net.cpp
index f7b6a2a333..227eec722f 100644
--- a/src/rpc/net.cpp
+++ b/src/rpc/net.cpp
@@ -956,7 +956,7 @@ static RPCHelpMan addpeeraddress()
         if (node.addrman->Add({address}, address)) {
             success = true;
             if (tried) {
-                // Mark entry as good and possibly move it to the "tried" addresses table.
+                // Attempt to move the address to the tried addresses table.
                 node.addrman->Good(address);
             }
         }
diff --git a/test/functional/rpc_net.py b/test/functional/rpc_net.py
index e2dfa3ea23..1b468aec22 100755
--- a/test/functional/rpc_net.py
+++ b/test/functional/rpc_net.py
@@ -254,16 +254,24 @@ class NetTest(BitcoinTestFramework):
 
         self.log.debug("Test that adding a valid address succeeds")
         assert_equal(node.addpeeraddress(address="1.2.3.4", port=8333), {"success": True})
-        addrs = node.getnodeaddresses(count=0)
-        assert_equal(len(addrs), 1)
-        assert_equal(addrs[0]["address"], "1.2.3.4")
-        assert_equal(addrs[0]["port"], 8333)
+        with node.assert_debug_log(expected_msgs=["Addrman checks started: new 1, tried 0, total 1"]):
+            addrs = node.getnodeaddresses(count=0)
+            assert_equal(len(addrs), 1)
+            assert_equal(addrs[0]["address"], "1.2.3.4")
+            assert_equal(addrs[0]["port"], 8333)
 
         self.log.debug("Test that adding the same address again when already present fails")
         assert_equal(node.addpeeraddress(address="1.2.3.4", port=8333), {"success": False})
         assert_equal(len(node.getnodeaddresses(count=0)), 1)
 
-        self.log.debug("Test adding a second address (with a different /16) to the tried table")
+        # RPC addpeeraddress sets the source address equal to the destination address.
+        # If addresses with the same /16 are passed to it, they will be placed in the
+        # same new bucket and have a 1/64 chance of the bucket positions colliding
+        # (depending on the value of nKey in the addrman), in which case the new
+        # address won't be added to the addrman.  Therefore, reduce the chance of a
+        # collision by using an address from a different /16.  There may still remain
+        # a 1/2^16 = 1/65536 probability of bucket/position collisions.
+        self.log.debug("Test adding a second address to the tried table")
         assert_equal(node.addpeeraddress(address="2.0.0.0", port=8333, tried=True), {"success": True})
         with node.assert_debug_log(expected_msgs=["Addrman checks started: new 1, tried 1, total 2"]):
             addrs = node.getnodeaddresses(count=0)  # getnodeaddresses re-runs the addrman checks
```

Reply:
(an alternative could be to merge the first valid address test with the "tried" test)

Reply:
or maybe first add an addr to tried, and the next one to new - then new collisions wouldn't be an issue.

Reply:
@mzumsande done

Reply:
Hmm, unfortunately this issue also seems to be possible in the regression test commit where 4 adds are added.

Reply:
Good catch -- inversed the order.